### ISSUE TITLE
Normalize Drugs

### ISSUE THREAD

**Issue by bwalsh** [ohsu-comp-bio/g2p-aggregator#8]
As a researcher, in order to evaluate evidence from different sources, I need to see a uniform drug name and identifier.

**Comment by bwalsh** [comment]
###  proposal


```
def normalize(name):
    """ call pubchem and retrieve compound_id and most common synonym
        see https://pubchem.ncbi.nlm.nih.gov/rdf/#_Toc421254632
    """
    url = 'https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/name/{}/synonyms/JSON'.format(name)  # NOQA
    r = requests.get(url)
    rsp = r.json()
    if 'InformationList' in rsp:
        informationList = r.json()['InformationList']
        information = informationList['Information'][0]
        return {'ontology_term': 'compound:CID{}'.format(information['CID']),
                'synonym': information['Synonym'][0]}
    else:
        return None

```

###  behavior

```
def test_bayer():
    info = normalize('Bayer')
    assert info
    assert info['ontology_term'] == 'compound:CID2244'
    assert info['synonym'] == 'aspirin'
```

**Comment by jgoecks** [comment]
@bwalsh Could we use http://c.biothings.io/ for normalization? "Drug and Compound API provides simple-to-use REST web services to query/retrieve drug annotation information. It integrates data from multiple resources, including chebi, chembl, drugbank, ginas, ndc, pharmgkb, pubchem, sider, unii, etc." It also includes RxNorm.

**Comment by bwalsh** [comment]
@jgoecks Can you double check?  The website endpoint seems a bit flakey... 
* c.biothings.io/v1/query?q=drugbank.name%3Aasprin  returns no hits.
* c.biothings.io/v1/query?q=drugbank.name%3Acelecoxib took over 30 secs returns a "boatload" of data

The curl interface seems to be stable and performant, if I trim the payload.
I'm a little concerned that neither "asprin" or "Acetylsalicylic Acid" return anything.  Perhaps there is a different query instead of drugbank.name ?


If we do use the endpoint I'd query it like this and return the FDA name
{ "ontology_term":"compound:CID2662" , "synonym":"Celecoxib"}

````
curl 'c.biothings.io/v1/query?q=drugbank.name%3Acelecoxib&fields=pubchem.cid,chembl.molecule_synonyms'
{
  "max_score": 16.30571,
  "took": 30,
  "total": 1,
  "hits": [
    {
      "_id": "RZEKVGVHFLEQIL-UHFFFAOYSA-N",
      "_score": 16.30571,
      "chembl": {
        "molecule_synonyms": [
          {
            "syn_type": "TRADE_NAME",
            "synonyms": "Celebrex"
          },
          {
            "syn_type": "RESEARCH_CODE",
            "synonyms": "SC-58635"
          },
          {
            "syn_type": "BAN",
            "synonyms": "Celecoxib"
          },
          {
            "syn_type": "FDA",
            "synonyms": "Celecoxib"
          },
          {
            "syn_type": "INN",
            "synonyms": "Celecoxib"
          },
          {
            "syn_type": "USAN",
            "synonyms": "Celecoxib"
          },
          {
            "syn_type": "TRADE_NAME",
            "synonyms": "Celebra"
          },
          {
            "syn_type": "TRADE_NAME",
            "synonyms": "Onsenal"
          }
        ]
      },
      "pubchem": {
        "cid": "CID2662"
      }
    }
  ]
}
```

**Comment by jgoecks** [comment]
@bwalsh "Acetylsalicylic acid" worked for me; "aspirin" did not: 
http://c.biothings.io/v1/query?q=drugbank.name:Acetylsalicylic%20acid

We should definitely experiment with the best querying approach, but the nice aspect of this store is that it includes results from other databases like SIDER (side effects database), which we'll want in the long run anyways.

**Comment by mayfielg** [comment]
It seems to be the case that when using the biothings frame on top of ES, at least some of the free text search capabilities of ES is lost. 

The following requests correctly grabs the 'aspirin' entry by the 'synonym' entries:
http://c.biothings.io/v1/query?q=chembl.molecule_synonyms.synonyms:aspirin
http://c.biothings.io/v1/query?q=chebi.synonyms:Acetol

Also not convinced that the API is 'flaky,' but rather that its a heavy search through a large database, if what you mean by flaky is a bit slow. Have you experienced problems other than slowness?

**Comment by bwalsh** [comment]
@jgoecks @mayfielg thanks for the follow up.

> Have you experienced problems other than slowness?
The swagger website kept crashing in chrome.

So, all things being equal, I'd query it like this ...
```
http://c.biothings.io/v1/query?q=chembl.molecule_synonyms.synonyms:aspirin&fields=pubchem.cid,chembl.molecule_synonyms
```

**Comment by bwalsh** [comment]
add command line options ...

**Comment by bwalsh** [comment]
@ahwagner Can you add your insights on the drug database you mentioned to this thread?

**Comment by ahwagner** [comment]
Yes! After reviewing Pubchem, ChEMBL, and DrugBank, I feel that ChEMBL is the best tool for normalizing drug concepts. They have 1.7 million molecules in the database, and include large molecules (e.g. biotherapeutics) which are not represented in PubChem compounds.

**Comment by bwalsh** [comment]
I have a work in progress that includes ChEMBL to the harvesters.

**Comment by bwalsh** [comment]
implemented in v0.3

**Comment by bwalsh** [comment]
![image](https://user-images.githubusercontent.com/47808/27341575-83aeb6e6-5592-11e7-9f04-458158c5e8e8.png)
Currently there are 9,868 pieces of ‘completely normalized data’  out of a total dataset of 15,974
